controllers/function_controller.go

```diff
@@ -0,0 +1,211 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/go-logr/logr"
+	openfunction "github.com/openfunction/apis/core/v1beta1"
+	"github.com/openfunction/pkg/util"
+	"github.com/openfunction/revision-controller/pkg/constants"
+	"github.com/openfunction/revision-controller/pkg/revision"
+	"github.com/openfunction/revision-controller/pkg/revision/git"
+	"github.com/openfunction/revision-controller/pkg/revision/image"
+	"github.com/openfunction/revision-controller/pkg/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	revisionKey       = "openfunction.io/revision-trigger"
+	revisionParamsKey = "openfunction.io/revision-trigger-params"
+)
+
+var (
+	commitShaRegEx = regexp.MustCompile(`^[0-9a-f]{7,40}$`)
+)
+
+// FunctionReconciler reconciles a Function object
+type FunctionReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	ctx    context.Context
+
+	revisions map[string]revision.Revision
+}
+
+func NewFunctionReconciler(mgr manager.Manager) *FunctionReconciler {
+	r := &FunctionReconciler{
+		Client:    mgr.GetClient(),
+		Scheme:    mgr.GetScheme(),
+		Log:       ctrl.Log.WithName("controllers").WithName("Function"),
+		revisions: make(map[string]revision.Revision),
+	}
+
+	return r
+}
+
+//+kubebuilder:rbac:groups=core.openfunction.io,resources=functions,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core.openfunction.io,resources=functions/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// the Function object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *FunctionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	log := r.Log.WithValues("Function", req.NamespacedName)
+
+	fn := &openfunction.Function{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+
+	if err := r.Get(ctx, req.NamespacedName, fn); err != nil {
+		if util.IsNotFound(err) {
+			log.V(1).Info("Function deleted")
+			r.cleanRevisionByFunction(fn)
+		}
+
+		return ctrl.Result{}, util.IgnoreNotFound(err)
+	}
+
+	if fn.Annotations == nil ||
+		fn.Annotations[revisionKey] != "enable" {
+		r.cleanRevisionByFunction(fn)
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, r.addRevision(fn, log)
+}
+
+func (r *FunctionReconciler) addRevision(fn *openfunction.Function, log logr.Logger) error {
```

Comment:
There is log field in r, do we still need to pass log here?
![image](https://user-images.githubusercontent.com/18525465/212795151-72dce3d3-01c1-4a6a-908c-d63b786efb3e.png)